pkg/mimirpb/custom.go

```diff
@@ -8,8 +8,76 @@ import (
 	"math"
 
 	"github.com/prometheus/prometheus/model/histogram"
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/encoding/proto"
+	"google.golang.org/grpc/mem"
+	protobufproto "google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/protoadapt"
 )
 
+func init() {
+	c := encoding.GetCodecV2(proto.Name)
+	encoding.RegisterCodecV2(&CodecV2{c})
+}
+
+// CodecV2 customizes gRPC unmarshalling.
+type CodecV2 struct {
+	encoding.CodecV2
```

Comment:
I would not embed this here, but convert it to a field and implement the two methods missing.

Otherwise it's unclear what are the other things that we're inheriting from the orignal codec: for example here this means that our codec has the same `Name()` as the parent: is that ok?

Reply:
It's definitely OK that `Name()` is the same as `encoding.CodecV2`, as it should replace it:

```golang
func init() {
	c := encoding.GetCodecV2(proto.Name)
	encoding.RegisterCodecV2(&CodecV2{c})
}
```

I think that embedding is good, because our implementation is supposed to only customize the `Unmarshal` method of the standard `CodecV2`.

Reply:
> I think that embedding is good, because our implementation is supposed to only customize the Unmarshal method of the standard CodecV2.

We can still "forward" implementation of other methods to original implementation, if we use named field. It would just be more explicit.

Reply:
Yes of course, it's possible to do the same without embedding. Help me understand though, why should we not use the embedding technique in this case? Are you against it in principle?

Reply:
> If this case isn't right for struct embedding, then I wonder which is?

I think struct embedding is valid in codebases that respect backwards incompatible changes. Given that the grpc package makes breaking changes in minor versions, it would be very useful, IMO, to detect that they suddenly added a new method to the CodecV2 interface that changes the behaviour in some drastic way.

> We can still "forward" implementation of other methods to original implementation, if we use named field. It would just be more explicit.

Yes, that's what I meant.

```go
type CodecV2 struct {
    wrapped encoding.CodecV2
}

func (c *CodecV2) Name() string { return c.wrapped.Name() }
```

Etc.

Reply:
No longer embedding.